### PR TITLE
verify that we don't run a session within another session

### PR DESCRIPTION
**Pull request by eishay**
following @2is10 suggestion. I had some deja vu since I implemented a similar code at wealthfront few years ago :-)
after adding it some tests failed (surprise!) since either the tests or the prod code opened a session within another session (bad bad).
@gmethvin @ymatsuda please review

### CODE REVIEW

File: shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala [hunk 1]
```diff
@@ -26,6 +27,18 @@ trait DataBaseComponent {
   def entityName(name: String): String = name
 }
 
+class InSessionException(message: String) extends Exception(message)
+
+object DatabaseSessionLock {
+  private val inSession = new DynamicVariable[Boolean](false)
+  def enteringSession[T](f: => T) = {
+    if (DatabaseSessionLock.inSession.value) throw new InSessionException("already in a DB session!")
+    val res = inSession.withValue(true) { f }
```
Comment (2is10) [inline shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala]: oooh, fancy!

File: shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala [hunk 2]
```diff
@@ -26,6 +27,18 @@ trait DataBaseComponent {
...
+    if (DatabaseSessionLock.inSession.value) throw new IllegalStateException("expected inSession to return false")
```
Comment (2is10) [inline shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala]: This line seems to imply mistrust of `DynamicVariable`. I'm fine with it for now, but we might eventually want to remove it.

**Comment by 2is10** [comment]
Fantastic. Thanks!

File: shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala (hunk 2 above)
Comment (eishay) [inline shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala]: I'll remove it now, tests show's we can trust it.

File: shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala (hunk 1 above)
Comment (gmethvin) [inline shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala]: do we need the extra val?

Comment (eishay) [inline shoebox/app/com/keepit/common/db/slick/DataBaseComponent.scala]: no, fixed

**Pull request merged**